lib/linalg.dx

```diff
@@ -4,33 +4,51 @@
 - Matrix Inversion
 
 
-def cast {a m} [Ix a, Ix m] (d:a) : m = (ordinal d)@m
+def unsafe_cast {a m} [Ix a, Ix m] (d:a) : m = unsafe_from_ordinal m (ordinal d)
+def unsafe_first_ix {n} [Ix n] : n = unsafe_from_ordinal n 0
+def unsafe_last_ix  {n} [Ix n] : n = unsafe_from_ordinal n (unsafe_nat_diff (size n) 1)
+
+def first_ix_of {n} [Ix n] (x:n) : n =
+  -- This function is safe because one can't produce an element of type n
+  -- if there isn't at least one possible value in its type. So being able
+  -- to produce x to call this function is proof that `n` is non-empty.
+  unsafe_from_ordinal n 0
+
+def last_ix_of {n} [NonEmpty n] (x:n) : n =
+  unsafe_from_ordinal _ $ unsafe_i_to_n $ n_to_i (size n) - 1
+
+def is_last {n} [Ix n] (x:n) : Bool =
+  ((ordinal x) + 1) == size n
+
 
 '### Triangular matrices
 
 def LowerTriMat (n:Type) [Ix n] (v:Type) : Type = i:n=>(..i)=>v
 def UpperTriMat (n:Type) [Ix n] (v:Type) : Type = i:n=>(i..)=>v
 
-def upper_tri_diag {n v} (u:UpperTriMat n v) : n=>v = for i. u.i.(0@_)
-def lower_tri_diag {n v} (l:LowerTriMat n v) : n=>v = for i. l.i.(cast i)
+def upper_tri_diag {n v} (u:UpperTriMat n v) : n=>v = view i. u.i.unsafe_first_ix
```

Comment:
Good idea.  It looks a little ugly since we have to pass a `..i`, but I think it's still an improvement:

```u.i.(first_ix_of (from_just $ project _ i))```

Fixed.